proxy/coco/src/peer/run_state.rs

```diff
@@ -0,0 +1,383 @@
+#[derive(Clone, Debug)]
+pub enum SyncEvent {
+    /// A sync has been initiated for `PeerId`.
+    Started(PeerId),
+    /// A sync has failed for `PeerId`.
+    Failed(PeerId),
+    /// A sync has succeeded for `PeerId`.
+    Succeeded(PeerId),
+}
+
+/// Scheduled timeouts which can occur.
+#[derive(Clone, Debug)]
+pub enum TimeoutEvent {
+    /// Grace period is over signaling that we should go offline, no matter how many syncs have
+    /// succeeded.
+    SyncPeriod,
+}
+
+/// The current status of the local peer and its relation to the network.
+#[derive(Debug)]
+enum Status {
+    /// Nothing is setup, not even a socket to listen on.
+    Stopped(Instant),
+    /// Local peer is listening on a socket but has not connected to any peers yet.
+    Started(Instant),
+    /// The local peer lost its connections to all its peers.
+    Offline(Instant),
+    /// Phase where the local peer tries get up-to-date.
+    Syncing(Instant, usize),
+    /// The local peer is operational and is able to interact with the peers it has connected to.
+    Online(Instant),
+}
+
+/// Set of knobs to change the behaviour of the [`RunState`].
+pub struct Config {
+    /// Number of peers a full sync is attempting with up on startup.
+    pub sync_max_peers: usize,
+    /// Enables the syncing stage when coming online.
+    pub sync_on_startup: bool,
+    /// Duration until the local peer goes online regardless if and how many syncs have succeeded.
+    pub sync_period: Duration,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            sync_max_peers: DEFAULT_SYNC_MAX_PEERS,
+            sync_on_startup: false,
+            sync_period: DEFAULT_SYNC_PERIOD,
+        }
+    }
+}
+
+/// State kept for a running local peer.
+pub struct RunState {
+    /// Confiugration to change how input [`Event`]s are interpreted.
+    config: Config,
+    /// Tracking remote peers that have an active connection.
+    connected_peers: HashSet<PeerId>,
+    /// Current internal status.
+    status: Status,
+}
+
+impl From<Config> for RunState {
+    fn from(config: Config) -> Self {
+        Self {
+            config,
+            connected_peers: HashSet::new(),
+            status: Status::Stopped(Instant::now()),
+        }
+    }
+}
+
+impl RunState {
+    /// Constructs a new state.
+    #[cfg(test)]
+    const fn new(config: Config, connected_peers: HashSet<PeerId>, status: Status) -> Self {
+        Self {
+            config,
+            connected_peers,
+            status,
+        }
+    }
+
+    /// Applies the `event` and based on the current state transforms to the new state and in some
+    /// cases produes commands which should be executed in the appropriate sub-routines.
+    pub fn transition(&mut self, event: Event) -> Vec<Command> {
+        log::trace!("run_state: ({:?}, {:?})", self.status, event);
+
+        match (&self.status, event) {
+            // Go from [`Input::Stopped`] to [`Input::Offline`] once we are listening.
```

Comment:
Not sure if I get this comment. Would this be correct?

```suggestion
            // Go from [`Status::Stopped`] to [`Status::Started`] once we are listening.
```

If yes, not sure the comment is necessary. but I might not be understanding things here